test/gtest/ucp/test_ucp_request.cc

```diff
@@ -247,142 +233,143 @@ class test_proto_reset : public ucp_test {
-                wait_for_condition(
-                        [this, i]() { return m_rbufs[i] == m_sbufs[i]; });
+        if (op == STREAM) {
+            for (unsigned i = 0; i < reqs.size(); ++i) {
+                get_stream_data(m_rbufs[i].get());
             }
+        } else if (op == AM) {
+            wait_for_value(&m_am_cb_cnt, reqs.size());
         }
 
         requests_wait(reqs);
-        flush_ep(sender());
+    }
+
+    void send_requests(unsigned reqs_count, std::vector<void*> &reqs,
+                       operation_e op, bool sync)
+    {
+        reqs.clear();
+        m_am_cb_cnt = 0;
 
         for (int i = 0; i < reqs_count; ++i) {
-            EXPECT_EQ(m_sbufs[i], m_rbufs[i]);
+            send_nb(m_sbufs[i], m_rbufs[i].get(), op, sync, reqs);
         }
     }
 
-    void init_rkey(std::vector<char> &sbuf)
+    void reset_protocol(operation_e op, bool sync = false)
     {
-        const ucp_mem_map_params_t params = {
-            .field_mask = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
-                          UCP_MEM_MAP_PARAM_FIELD_LENGTH,
-            .address    = sbuf.data(),
-            .length     = sbuf.size()
-        };
-
-        ucp_mem_h memh;
-        ASSERT_UCS_OK(ucp_mem_map(receiver().ucph(), &params, &memh));
-        m_memhs.push_back(memh);
-
-        void *rkey_buffer;
-        size_t rkey_buffer_size;
-        ASSERT_UCS_OK(ucp_rkey_pack(receiver().ucph(), memh, &rkey_buffer,
-                                    &rkey_buffer_size));
-
-        ucp_rkey_h rkey;
-        ASSERT_UCS_OK(ucp_ep_rkey_unpack(sender().ep(), rkey_buffer, &rkey));
-        ucp_rkey_buffer_release(rkey_buffer);
-        m_rkeys.push_back(rkey);
-    }
+        static const unsigned reqs_count = 1000;
+        static const size_t msg_size     = UCS_KBYTE * 64;
+
+        sender().connect(&receiver(), get_ep_params());
+        receiver().connect(&sender(), get_ep_params());
 
-    typedef std::function<bool()> predicate_t;
+        for (int i = 0; i < reqs_count; ++i) {
+            mapped_buffer *rbuf = new mapped_buffer(msg_size, receiver());
+            rbuf->memset(0);
+            m_rbufs.push_back(std::unique_ptr<mapped_buffer>(rbuf));
+            m_sbufs.push_back(std::vector<uint8_t>(msg_size));
+        }
 
-    void wait_for_condition(const predicate_t &predicate)
-    {
-        const double timeout      = 10;
-        const ucs_time_t deadline = ucs::get_deadline(timeout);
+        /* Send a single message to complete wireup before sending actual
+           data */
+        std::vector<void*> reqs;
+        send_requests(1, reqs, op, sync);
+        wait_recv(op, reqs);
 
-        while (!predicate() && (ucs_get_time() < deadline)) {
-            progress();
+        /* Send all messages */
+        send_requests(reqs_count, reqs, op, sync);
+        restart(sender().ep());
+
+        wait_recv(op, reqs);
+
+        flush_ep(sender());
+
+        for (int i = 0; i < reqs_count; ++i) {
+            EXPECT_EQ(0,
```

Comment:
can we do
`EXPECT_EQ(m_sbufs[i], m_rbufs[i])`

Reply:
those are different types - sbuf is a vector and rbuf is mapped_buffer